### ISSUE TITLE
DropoutWrapper and Exploding gradient behaviour for Recurrent Neural Network

### ISSUE THREAD

**Issue by tanodino** [tensorflow/tensorflow#35181]
Dear all,
 I have a point about DropoutWrapper and its use with Recurrent Neural Networks.

Due to the possibility that the dropout can be applied to the state or the output (state_keep_prob and output_keep_prob), I found that, during the recurrent process, the state propagated through the time can take values not bounded in the interval [-1, 1]. This is probably due to the way in which the dropout is implemented (at training time with a scaling instead of testing time with expectation). Since the dropout is applied after the activation (i.e. tanh), the feature values will range between -inf and +inf. This point is a bit strange for me since the current implementation can induce exploding gradient issues in the GRU/LSTM process while such cells were introduced to deal with vanishing as well as exploding gradients.

Please, could you supply me some feedback about my issue since, practically, it can impact people that commonly employ such Wrapper that induces behaviours that are divergent w.r.t. the theoretical behaviour of RNN (GRU/LSTM).

All the best

**Comment by gadagashwini** [comment]
@tanodino, Could you provide the sample code snippet to analyze the reported issue. Thanks!

**Comment by tanodino** [comment]
Hi, I wrote some code example on a notebook that you can run on colab.
Here the link to the notebook:
https://colab.research.google.com/drive/1rNkHlySCXXasRqf3sCbYoSGSK5BdWT2h

In this code, I generated some random data and run a GRUCell wrapped by a DropoutWrapper layer. During the training I collected the value of the final state of the RNN and print them. As you can observe, such values are not bounded in the interval [-1, 1] and this can "theoretically" impact the gradient propagation since it can induces exploding gradient if the number of timestamps will be high. On the other hand, if you comment the line:

cell = DropoutWrapper(cell, output_keep_prob=1.0-dropOut, state_keep_prob=1.0-dropOut)

you will see that the value of the "feat" tensor will be bounded in the interval [-1, 1] respecting the theoretical consideration related to GRU and LSTM cells.

I hope to haver well highlighted the point I want to raise up.

Best

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
